### PR TITLE
f/schedule update endpoint

### PR DESCRIPTION
**Pull request by nataliemona**

### CODE REVIEW

File: functions/schedule-update.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    TODO: Explain the purpose of this function.
+    """
+    sheet = GSheet(GSHEETS_DOC_ID, in_lambda=True)
+    sheet.import_to_dynamo(GSHEETS_SHEET_NAME)
+
+    return {
+        "statusCode": 200,
+        "headers": {},
+        "body": "Success"
+    }
```
Comment (nataliemona) [inline functions/schedule-update.py]: missing new line - I'll fix my editor

File: functions/schedule-update.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+GSHEETS_DOC_ID = env("GSHEETS_DOC_ID")
+GSHEETS_SHEET_NAME = env("GSHEETS_SHEET_NAME")
+
+logger = logging.getLogger()
+
+
+@setup_logging
+def handler(event, context):
+    """
+    TODO: Explain the purpose of this function.
```
Comment (nataliemona) [inline functions/schedule-update.py]: remove this or fix this

File: functions/gsheets.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        Save credentials file to SSM parameter.
+        """
+        if not filename:
+            filename = self.default_filename
+        if not exists(filename):
+            raise Exception(f"Credentials file {filename} not found")
+
+        data = ""
+        with open(filename, "r") as f:
+            data = json.dumps(f.read())
```
Comment (lbjay) [inline functions/gsheets.py]: This doesn't seem right. I would expect a `json.load(f)` here.

File: cdk/stack.py [hunk 4]
```diff
@@ -66,6 +71,24 @@ def __init__(self, scope: core.Construct, id: str,
...
+            lambda_code_bucket=lambda_code_bucket,
+            environment={
+                "CLASS_SCHEDULE_TABLE": schedule.table.table_name,
+                "GSHEETS_DOC_ID": gsheets_doc_id,
+                "GSHEETS_SHEET_NAME": gsheets_sheet_name,
+            }
+        )
+        schedule_update.function.add_to_role_policy(
+            iam.PolicyStatement(
+                actions=["ssm:GetParameter", "ssm:PutParameter"],
```
Comment (lbjay) [inline cdk/stack.py:85]: does the function ever need to write to parameter store?

File: functions/gsheets.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+from googleapiclient.discovery import build
+
+SCHEDULE_TABLE = env("CLASS_SCHEDULE_TABLE")
+
+logger = logging.getLogger()
+
+load_dotenv(join(dirname(__file__), '../.env'))
+
+
+class GSheetsAuth():
```
Comment (lbjay) [inline functions/gsheets.py:25]: It's maybe not worth redoing, but something to think about in the future re: putting stuff in classes. I'm not sure it's warranted in this situation. Usually you would do that if you wanted to have functions (methods) that shared some kind of state between calls. Here it seems like any given process is only going to loading, saving or deleting once.

File: functions/gsheets.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+        r = self.ssm.put_parameter(
+            Name=self.ssm_path,
+            Description="Service account creds for gsheets authorization",
+            Value=json.loads(data),
+            Type="SecureString",
+            Overwrite=True,
+        )
+        logger.info(f"Saved token version {r['Version']}")
+
+    def delete_from_ssm(self):
```
Comment (lbjay) [inline functions/gsheets.py:78]: does this get run on stack.delete?

File: tasks.py [hunk 7]
```diff
@@ -590,96 +601,18 @@ def get_day_of_week_code(day):
             raise Exit("Failed converting to dynamo item format: {}\n{}" \
                        .format(str(e), json.dumps(event, indent=2)))
 
-    __schedule_json_to_dynamo(ctx, schedule_data=schedule_data)
+    schedule_json_to_dynamo(
```
Comment (lbjay) [inline tasks.py]: this is calling `schedule_json_to_dyanmo` with an invalid arg set. do we still need this import from opencast task?

File: functions/gsheets.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+        logger.debug(f"Successfully downloaded {sheet_name}.csv")
+        return file_path
+
+    def import_to_dynamo(self, sheet_name):
+        file_path = self._download_csv(sheet_name)
+
+        schedule_csv_to_dynamo(SCHEDULE_TABLE, file_path)
+
+
+def schedule_json_to_dynamo(schedule_table, json_file=None, schedule_data=None):
```
Comment (lbjay) [inline functions/gsheets.py]: not asking for changes; just thinking out loud. i wrote this code iirc and it feels a bit inelegant and possibly error prone. I was fine with that when it only ran locally and i could play with it and comment out stuff. Now that it's going to run in a lambda that has to be packaged, deployed, etc, I wonder if we're going to wish it was broken up a bit and more testable. IDK, what do you think?

**Review by lbjay — CHANGES_REQUESTED** [review]

File: cdk/stack.py (hunk 4 above)
Comment (nataliemona) [reply cdk/stack.py:85]: No

File: functions/gsheets.py (hunk 3 above)
Comment (nataliemona) [reply functions/gsheets.py]: Thanks :-)

File: functions/gsheets.py (hunk 6 above)
Comment (nataliemona) [reply functions/gsheets.py:78]: No... maybe we should just remove this since credentials are being shared.

File: tasks.py (hunk 7 above)
Comment (nataliemona) [reply tasks.py]: Good point. Let's keep the import form opencast code though until we're sure we won't be needing it anymore?

File: functions/gsheets.py (hunk 8 above)
Comment (nataliemona) [reply functions/gsheets.py]: Added a unit test for the parsing code - should add more as well.

File: tasks.py (hunk 7 above)
Comment (lbjay) [reply tasks.py]: That's fine. It's always in the git history though, don't forget.

**Pull request merged**